### DISCUSSION TITLE
howto: update helm chart without dependencies

### DISCUSSION THREAD

**Discussion by Fagioletti** [renovatebot/renovate#21790] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.35.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

14.10.5-ee

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

PR should contain:
- new Chart.yaml
- new values.yaml
- new resources (files) in /template

### What you tried so far.

- read documentation
- worked through discussions-threads
- fiddled around with Chart.yaml, Regex Manager
- tried to find out in source-code (of renovate) how to get this to work.

a lot of applications have 'simple' Chart.yaml(s) - see example below. 
i could not find a way to get renovate suggest updates to 'plain' Chart.yaml's which do not have dependencies. Do i have to change those to declare themselves as dependencies to get them updated (in charts)?
what are best practices to get such apps updated?

Example: stakater/reloader
``` yaml
apiVersion: v2
name: reloader
version: v0.0.79
appVersion: v0.0.79
description: Reloader chart that runs on kubernetes
home: https://github.com/stakater/Reloader
icon: https://raw.githubusercontent.com/stakater/Reloader/master/assets/web/reloader-round-100px.png
keywords:
- Reloader
- kubernetes
maintainers:
- email: hello@stakater.com
  name: Stakater
sources:
- https://github.com/stakater/IngressMonitorController
```

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
What do you expect to be updated in that file?

**Reply by Fagioletti** [reply]
> What do you expect to be updated in that file?

- version
- appVersion

to rephrase (hopefully clearer) what the final goal would be:
i have reloader v0.0.79 in (see original Chart.yaml above, save apiVersion which originally is v1):

``` 
│   └── helm
│       └── reloader
│           ├── Chart.yaml
│           ├── templates
│           │   ├── NOTES.txt
│           │   ├── _helpers.tpl
│           │   ├── clusterrole.yaml
│           │   ├── clusterrolebinding.yaml
│           │   ├── deployment.yaml
│           │   ├── role.yaml
│           │   ├── rolebinding.yaml
│           │   ├── service.yaml
│           │   ├── serviceaccount.yaml
│           │   └── servicemonitor.yaml
│           └── values.yaml
```

i'd like that
- renovate detects a new version for reloader 
- gets the new version, downloads and untars for me to compare all changes in the PR with the current version

i found [Issue: Update helm tarball files, on helm chart repositories with dependencies #18110](https://github.com/renovatebot/renovate/issues/18110) or 
[Issue: upgrade appversion and chart version in chart.yaml #19048](https://github.com/renovatebot/renovate/discussions/19048) which seem part of the solution, but i can't get it together (in my head).

thanks for your input.

**Reply by secustor** [reply]
What you are looking for is currently not supported, but what you can do is follow this issue: https://github.com/renovatebot/renovate/issues/4759

Renovate it self does not change template files.  
What you can do is to create a "Meta" `Chart.yaml` with your charts as subchart and then make usage of `postUpgradeTask` to pull and commit the files.

**Reply by rarkins** [reply]
If there is a repo like `https://github.com/stakater/Reloader` in this Chart and the `appVersion` of `0.0.79` refers to it, it makes sense to me that our Helm charts manage can detect and update this. If updating it requires _files_ to then be pulled into the same repo, ideally that would be an "artifacts" capability instead of requiring a `postUpgradeTask` (assuming it's standardized)

**Reply by Fagioletti** [reply, marked as answer]
so based on @secustor's guidance i will:
- create (meta) Chart.yaml
- use [postUpdateOption helmUpdateSubChartArchives](https://github.com/renovatebot/renovate/pull/18162) to get the (new) helm chart as .tgz
- let [postUpgradeTasks](https://docs.renovatebot.com/configuration-options/#postupgradetasks) (un)tar, mv, cleanup

thanks a lot.

**Reply by SimeonPoot** [reply]
Do you have a concrete example of the usage of postUpgradeTasks.. I can't get it to work.. I'm testing this out with cert-manager helm charts.. Based on the proposed configuration above, I get a 'new' tarbal and the version is updated into the 'Meta' Chart.yaml file... But this isn't sufficient for my use-case.

I'd like to have the cert-manager (child) chart updated (e.g. all the manifests are updated).

The current setup: 
Within a directory: charts
I have a 'Meta' chart, which I add the cert-manager as dependency; 
Chart.yaml
```
...
apiVersion: v2
appVersion: "v1.10.2"
version: "v1.10.2"
dependencies:
- name: cert-manager
  version: "v1.10.2"  # "v1.13.0" is the newest 
  repository: https://charts.jetstack.io
...
```
next to this Chart.yaml i have another directory: 
charts/cert-manager
with the 'official/ upstream' (child) helm-chart of cert-manager untarred. 

So the directory structure from ./helm would be: 
```
.
├── Chart.yaml
├── charts
│   └── cert-manager
│       ├── Chart.yaml
│       ├── README.md
│       ├── templates
│       └── values.yaml
└── values.yaml
```
This is the renovate.json

`{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "platform": "github",
  "repositories": ["SimeonPoot/helm-charts"],
  "extends": [
   "config:recommended", ":dependencyDashboard"
  ],
  "postUpdateOptions": ["helmUpdateSubChartArchives"],
  "packageRules": [
    {
      "matchDatasources": ["gomod", "go", "helm", "golang-version", "helmv3"],
      "enabled": true,
      "prCreation": "immediate"
    }
  ],
  "allowedPostUpgradeCommands": [ "^install-tool helm v3.8.1$", "^helm dependency update$","^tar xz"],
  "postUpgradeTasks": {
    "commands": [ "install-tool helm v3.8.1", "helm dependency update", "tar xz" ],
    "fileFilters": [
      "Chart.yaml",
      "charts/*.tgz"
    ]
  }
 }
`

I tried to do a `postCommand` and untar it (tar xz ..), but I couldn't get Renovate to work with this (self-hosted cronjob from k8s). It didn't pick up the tar command. Also, it felt 'hacky'.

**Reply by viceice** [reply]
the allowed commands must be at bot config, not repo config.

**Reply by SimeonPoot** [reply]
hi @viceice, @Fagioletti, @secustor ... I've tried this, I understand that the two configs are being merged before a Renovate Run? 
The result of the config above seems to be the same: 
I'm trying this 'bot' config from kubernetes cronjob:
(image: renovatebot/renovate:36.87.7) 
```json
    {
      "platform": "github",
      "token": "xyz",
      "repositories": ["SimeonPoot/helm-charts"],
        "packageRules": [
        {
          "matchDatasources": ["gomod", "go", "helm", "golang-version", "helmv3"],
          "enabled": true,
          "prCreation": "immediate"
        }
      ],
      "postUpdateOptions": ["helmUpdateSubChartArchives"],
      "allowedPostUpgradeCommands": [ "^install-tool helm v3.8.1$", "^helm dependency update$", "^tar xf"],
      "postUpgradeTasks": {
        "commands": ["install-tool helm v3.8.1", "helm dependency update", "tar xf"],
        "fileFilters": [
          "helm/Chart.yaml",
          "helm/charts/cert-manager-v1.*.0.tgz"
          ]
        },
    }
```
results into this: 
https://github.com/SimeonPoot/helm-charts/pull/15
I get 2 errors (see the **Artifact update problem** in the PR)... 

I only want cert-managers (child chart) to be updated with the correct version + **manifests**! I can't comprehend why this is so hard to do....
Would you mind helping me out here?

**Reply by rarkins** [reply]
perhaps it's being run in the wrong directory?

**Comment by SimeonPoot** [comment]
this can be the case. I don't graps how to change to the correct directory jet.. if you have some insights in this, please advise. 

I hard-coded the path .tgz and where to untar (-C), to see if it can be untarred: 

```json
    {
      "platform": "github",
      "token": "xyz",
      "repositories": ["SimeonPoot/helm-charts"],
        "packageRules": [
        {
          "matchDatasources": ["gomod", "go", "helm", "golang-version", "helmv3"],
          "enabled": true,
          "prCreation": "immediate"
        }
      ],
      "postUpdateOptions": ["helmUpdateSubChartArchives"],
      "allowedPostUpgradeCommands": [ "^install-tool helm v3.8.1$", "^helm dependency update$", "^tar xf /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts/cert-manager-v1.13.0.tgz -C /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts/$"],
      "postUpgradeTasks": {
        "commands": ["install-tool helm v3.8.1", "helm dependency update", "tar xf /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts/cert-manager-v1.13.0.tgz -C /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts/"],
        "fileFilters": [
          "helm/Chart.yaml",
          "SimeonPoot/helm-charts/helm/cert-manager-v1.*.0.tgz"
          ]
        },
    }
```
Checked at runtime if the file was present, and it was:
```bash
k exec -t renovate-28262548-rpcpp  -- ls /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts
cert-manager
cert-manager-v1.13.0.tgz

--- container logs:
DEBUG: Executing command (repository=SimeonPoot/helm-charts, branch=renovate/cert-manager-1.x)
       "dep": "cert-manager",
       "command": "tar xf /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts/cert-manager-v1.13.0.tgz -C /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts/"
DEBUG: exec completed (repository=SimeonPoot/helm-charts, branch=renovate/cert-manager-1.x)
       "dep": "cert-manager",
       "durationMs": 22,
       "stdout": "",
       "stderr": ""
DEBUG: Executed post-upgrade task (repository=SimeonPoot/helm-charts, branch=renovate/cert-manager-1.x)
       "dep": "cert-manager",
       "cmd": "tar xf /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts/cert-manager-v1.13.0.tgz -C /tmp/renovate/repos/github/SimeonPoot/helm-charts/helm/charts/",
       "stderr": "",
       "stdout": ""
```

Do you have any (other?) solution for this?

**Comment by ira-gordin-sap** [comment]
@SimeonPoot did you find solution?

**Reply by SimeonPoot** [reply]
@ira-gordin-sap, unfortunately I haven't found a solution for this.. After a couple of hours grinding, I changed my current setup, which I'd rather not have. I'll probably move away from Renovate and develop something custom for this use-case.
